circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/PausableState.kt

```diff
@@ -58,18 +57,40 @@ public fun <T> pausableState(
   isActive: Boolean = LocalRecordLifecycle.current.isActive,
   produceState: @Composable () -> T,
 ): T {
-  var state: T? by remember(key) { mutableStateOf(null) }
-
+  val pauseableState = remember(key) { PauseableState<T>() }.also { it.update(produceState) }
   val saveableStateHolder = rememberSaveableStateHolder()
 
-  if (isActive || state == null) {
+  if (isActive || pauseableState.value == null) {
     val retainedStateRegistry = rememberRetained(key = key) { RetainedStateRegistry() }
     CompositionLocalProvider(LocalRetainedStateRegistry provides retainedStateRegistry) {
       saveableStateHolder.SaveableStateProvider(key = key ?: "pausable_state") {
-        state = produceState()
+        pauseableState.Produce()
       }
     }
   }
 
-  return state!!
+  return pauseableState.value!!
+}
+
+/**
+ * Wrapper class to update state from inside the [CompositionLocalProvider] without the
+ * [MutableState] observed in such a way that the back write will trigger a recomposition.
+ */
+private class PauseableState<T> : State<T> {
+
+  private var produceState: (@Composable () -> T)? = null
+
+  private val state = mutableStateOf<T?>(null)
+
+  override val value: T
+    get() = state.value!!
```

Comment:
there's a lot of `!!` calls that incur runtime nullchecks that we know won't happen. Should we port over our `sneakyNull`? 

https://github.com/slackhq/slack-gradle-plugin/blob/f65322f31d99753b10511cf13c3a1d39d176a1b6/slack-plugin/src/main/kotlin/slack/gradle/util/PropertyUtil.kt#L29-L43